src/gate/parser/bench/parser.cpp

```diff
@@ -0,0 +1,290 @@
+    }
+  }
+}
+
+void doubleDefinition(Tokens whereCalled, std::size_t &line, std::vector<TokenMap>& maps) {
+  for (auto i : maps) {
+    if (i.definite == 1) { // definite = {0 if not declared, 1 if declared}.
+      if (strlen(yytext) == i.name.size() && whereCalled != TOK_OUTPUT
+        && i.typeInit != 2)  {
+        COMPARE("DOUBLE DEFINITION", line);
+        } else if (strlen(yytext) == i.name.size() 
+          && whereCalled == TOK_OUTPUT ) {
+          if(i.typeInit == 2)
+            COMPARE("DOUBLE DEFINITION", line);
+        }
+    }
+  }
+}
+
+Tokens getNextToken(std::size_t &place, std::size_t &line) { 
+  Tokens val = static_cast<Tokens>(scan_token());
+  std::cout << " token: " << val << "\t" << yytext << "\tline: " 
+    << line << " \tplace: " << place << std::endl;
+  place += 1;
+  return val;
+}
+
+void assertNextId(Tokens expectedToken,
+                  const char* errorString, 
+                  Tokens whereCalled,
+                  std::size_t &place,
+                  std::size_t &line,
+                  std::vector<TokenMap> &maps) {
+  Tokens token = getNextToken(place, line); 
+  if (token != expectedToken) { 
+    ALERT(errorString, line); 
+  }   
+  if (whereCalled == TOK_INPUT) { // checkers.
+    doubleDefinition(TOK_INPUT, line, maps);
+    MAPS(1, TOK_INPUT, maps);
+    // Gate::Id inOutId = net.addIn();
+  } else if (whereCalled == TOK_OUTPUT) {
+    doubleDefinition(TOK_OUTPUT, line, maps);
+    MAPS(1, TOK_OUTPUT, maps);
+  } else {
+    output(whereCalled, line, maps);
+    MAPS(0, whereCalled, maps);
+  } 
+}
+
+void parseParenthesisIO(Tokens type, 
+                        std::size_t &place,
+                        std::size_t &line,
+                        std::vector<TokenMap> &maps) { // type = {INPUT, OUTPUT, DFF, NOT}
+  ASSERT_NEXT_TOKEN(TOK_LP, "LP", place, line);
+  assertNextId(TOK_ID, "ID", type, place, line, maps);
+  ASSERT_NEXT_TOKEN(TOK_RP, "RP", place, line);
+  line += 1;
+}
+
+void parseParenthesisID(Tokens type, std::size_t &place, std::size_t &line, std::vector<TokenMap> &maps) { // type = {AND, OR, NAND, NOR}
+  Tokens token;
+  ASSERT_NEXT_TOKEN(TOK_LP, "LP", place, line);
+  assertNextId(TOK_ID, "ID", type, place, line, maps);
+  while ((token = getNextToken(place, line)) == TOK_COMMA)
+    assertNextId(TOK_ID, "ID", type, place, line, maps);
+  if (token != TOK_RP) {
+    ALERT("'RP'", line);
+  }
+  line += 1;
+}
+
+void parseID(std::size_t &place, std::size_t &line, std::vector<TokenMap>& maps) {
+  MAPS(1, TOK_E, maps); // adding.
+  ASSERT_NEXT_TOKEN(TOK_E, "E", place, line);
+  Tokens token = getNextToken(place, line);
+  if (token == TOK_AND || token == TOK_OR || token == TOK_NAND || token == TOK_NOR) {
+    maps.back().typeInit = token;
+    parseParenthesisID(token, place, line, maps);
+  } else if (token == TOK_DFF || token == TOK_NOT) { 
+    maps.back().typeInit = token;
+    parseParenthesisIO(token, place, line, maps);
```

Comment:
Are branches identical?

Reply:
No, difference in the function names "parseParenthesisI + D" and "parseParenthesisI + O". I will rename them then.